File: common/src/main/java/Point.java
import com.fasterxml.jackson.annotation.JsonProperty;


record Point(
        @JsonProperty("x") double x,
        @JsonProperty("y") double y
) implements Jsonisable {
    public static final Point ZERO = new Point(0.0, 0.0);

    public Point add(Point other) {
        return new Point(x + other.x(), y + other.y());
    }

    public Point diff(Point other) {
        return new Point(x - other.x(), y - other.y());
    }

    public Point mult(double scalar) {
        return new Point(x * scalar, y * scalar);
    }

    public Point div(double scalar) {
        return new Point(x / scalar, y/ scalar);
    }

    public double size() {
        return Math.sqrt(x * x + y * y);
    }

    public double dist(Point other) {
        return diff(other).size();
    }

    public Point normalize() {
        var size = size();
        if (size != 0.0) {
            return div(size);
        }
        return ZERO;
    }

    public Point rotate(double angle) {
        var cos = Math.cos(angle);
        var sin = Math.sin(angle);
        return new Point(x * cos - y * sin, x * sin + y * cos);
    }

    @Override
    public String toString() {
        return "Point{" +
                "x=" + x +
                ", y=" + y +
                '}';
    }
}

File: common/src/main/kotlin/demo/mesharena/common/Point.kt
package demo.mesharena.common

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import kotlin.math.cos
import kotlin.math.sin
import kotlin.math.sqrt

private val zero = Point(0.0, 0.0)

data class Point @JsonCreator constructor(
        @JsonProperty("x") val x: Double,
        @JsonProperty("y") val y: Double
) : Jsonisable {
  fun add(other: Point): Point {
    return Point(x + other.x, y + other.y)
  }

  fun diff(other: Point): Point {
    return Point(x - other.x, y - other.y)
  }

  fun mult(scalar: Double): Point {
    return Point(x * scalar, y * scalar)
  }

  operator fun div(scalar: Double): Point {
    return Point(x / scalar, y / scalar)
  }

  fun size(): Double {
    return sqrt(x * x + y * y)
  }

  fun normalize(): Point {
    val size = size()
    return if (size != 0.0) {
      div(size)
    } else zero
  }

  fun rotate(angle: Double): Point {
    val cos: Double = cos(angle)
    val sin: Double = sin(angle)
    return Point(x * cos - y * sin, x * sin + y * cos)
  }

  override fun toString(): String {
    return "Point{x=${x}, y=${y}}"
  }
}
